File: Serenity.Script.Core/Q/Q.ScriptData.cs
using jQueryApi;
using System;
using System.Collections.Generic;
using System.Html;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;

namespace Serenity
{
    public static partial class Q
    {
        public static class ScriptData
        {
            private static JsDictionary<string, string> registered = new JsDictionary<string, string>();
            private static JsDictionary<string, object> loadedData = new JsDictionary<string, object>();

            public static void BindToChange(string name, string regClass, Action onChange)
            {
                jQuery.FromObject(Document.Body).As<dynamic>()
                    .bind("scriptdatachange." + regClass, new Action<jQueryEvent, string>((e, s) =>
                {
                    if (s == name)
                        onChange();
                }));
            }

            public static void TriggerChange(string name)
            {
                J(Document.Body).TriggerHandler("scriptdatachange", new object[] { name });
            }

            public static void UnbindFromChange(string regClass)
            {
                jQuery.FromObject(Document.Body)
                    .Unbind("scriptdatachange." + regClass);
            }

            private static void SyncLoadScript(string url)
            {
                jQuery.Ajax(new jQueryAjaxOptions
                {
                    Async = false,
                    Cache = true,
                    Type = "GET",
                    Url = url,
                    Data = null,
                    DataType = "script"
                });
            }

            private static Promise LoadScriptAsync(string url)
            {
                return Promise.Void.ThenAwait(() => 
                {
                    Q.BlockUI();

                    return Promise.Resolve(jQuery.Ajax(new jQueryAjaxOptions
                    {
                        Async = true,
                        Cache = true,
                        Type = "GET",
                        Url = url,
                        Data = null,
                        DataType = "script"
                    }).Always(Q.BlockUndo));
                });
            }

            private static void LoadScriptData(string name)
            {
                if (!registered.ContainsKey(name))
                    throw new Exception(String.Format("Script data {0} is not found in registered script list!", name));

                name = name + ".js?" + registered[name];

                SyncLoadScript(Q.ResolveUrl("~/DynJS.axd/") + name);
            }

            private static Promise LoadScriptDataAsync(string name)
            {
                return Promise.Void.ThenAwait(() =>
                {
                    if (!registered.ContainsKey(name))
                        throw new Exception(String.Format("Script data {0} is not found in registered script list!", name));

                    name = name + ".js?" + registered[name];

                    return LoadScriptAsync(Q.ResolveUrl("~/DynJS.axd/") + name);
                });
            }

            public static object Ensure(string name)
            {
                var data = loadedData[name];

                if (!Script.IsValue(data))
                    LoadScriptData(name);

                data = loadedData[name];

                if (!Script.IsValue(data))
                    throw new NotSupportedException(String.Format("Can't load script data: {0}!", name));

                return data;
            }

            [IncludeGenericArguments(false)]
            public static Promise<TData> EnsureAsync<TData>(string name)
            {
                return Promise.Void.ThenAwait(() =>
                {
                    var data = loadedData[name].As<TData>();

                    if (Script.IsValue(data))
                        return Promise.FromValue(data);

                    return LoadScriptDataAsync(name).ThenSelect(() =>
                    {
                        data = loadedData[name].As<TData>();

                        if (!Script.IsValue(data))
                            throw new NotSupportedException(String.Format("Can't load script data: {0}!", name));

                        return data;
                    });
                });
            }

            public static object Reload(string name)
            {
                if (!registered.ContainsKey(name))
                    throw new NotSupportedException(String.Format("Script data {0} is not found in registered script list!"));

                registered[name] = new JsDate().GetTime().ToString();

                LoadScriptData(name);

                var data = loadedData[name];

                return data;
            }

            [IncludeGenericArguments(false)]
            public static Promise<TData> ReloadAsync<TData>(string name)
            {
                return Promise.Void.ThenAwait(() =>
                {
                    if (!registered.ContainsKey(name))
                        throw new NotSupportedException(String.Format("Script data {0} is not found in registered script list!"));

                    registered[name] = new JsDate().GetTime().ToString();

                    return LoadScriptDataAsync(name)
                        .ThenSelect(() => loadedData[name].As<TData>());
                });
            }

            public static bool CanLoad(string name)
            {
                var data = loadedData[name];

                if (Script.IsValue(data) ||
                    registered.ContainsKey(name))
                    return true;

                return false;
            }

            public static void SetRegisteredScripts(JsDictionary<string, string> scripts)
            {
                registered.Clear();
                foreach (var k in scripts)
                    Q.ScriptData.registered[k.Key] = k.Value.ToString();
            }

            public static void Set(string name, object value)
            {
                Q.ScriptData.loadedData[name] = value;
                TriggerChange(name);
            }
        }

        [IncludeGenericArguments(false)]
        public static TData GetRemoteData<TData>(string key)
        {
            return ScriptData.Ensure("RemoteData." + key).As<TData>();
        }

        [IncludeGenericArguments(false)]
        public static Promise<TData> GetRemoteDataAsync<TData>(string key)
        {
            return ScriptData.EnsureAsync<TData>("RemoteData." + key);
        }

        [IncludeGenericArguments(false)]
        public static Lookup<TItem> GetLookup<TItem>(string key)
        {
            return ScriptData.Ensure("Lookup." + key).As<Lookup<TItem>>();
        }

        [IncludeGenericArguments(false)]
        public static Promise<Lookup<TItem>> GetLookupAsync<TItem>(string key)
        {
            return ScriptData.EnsureAsync<Lookup<TItem>>("Lookup." + key);
        }

        public static void ReloadLookup(string key)
        {
            ScriptData.Reload("Lookup." + key);
        }

        public static Promise ReloadLookupAsync(string key)
        {
            return ScriptData.ReloadAsync<object>("Lookup." + key);
        }

        [IncludeGenericArguments(false)]
        public static List<PropertyItem> GetColumns(string key)
        {
            return ScriptData.Ensure("Columns." + key).As<List<PropertyItem>>();
        }

        public static Promise<List<PropertyItem>> GetColumnsAsync(string key)
        {
            return ScriptData.EnsureAsync<List<PropertyItem>>("Columns." + key);
        }

        [IncludeGenericArguments(false)]
        public static List<PropertyItem> GetForm(string key)
        {
            return ScriptData.Ensure("Form." + key).As<List<PropertyItem>>();
        }

        public static Promise<List<PropertyItem>> GetFormAsync(string key)
        {
            return ScriptData.EnsureAsync<List<PropertyItem>>("Form." + key);
        }

        public static string GetTemplate(string key)
        {
            return ScriptData.Ensure("Template." + key).As<string>();
        }

        public static Promise<string> GetTemplateAsync(string key)
        {
            return ScriptData.EnsureAsync<string>("Template." + key);
        }

        public static bool CanLoadScriptData(string name)
        {
            return ScriptData.CanLoad(name);
        }
    }
}

File: Serenity.Web/Scripts/serenity/CoreLib/Q/Q.ScriptData.ts
/// <reference path="Q.BlockUI.ts" />
/// <reference path="Q.StringHelpers.ts" />
/// <reference path="Q.Url.ts" />

namespace Q {

    export namespace ScriptData {
        let registered: { [key: string]: any } = {};
        let loadedData: { [key: string]: any } = {};

        export function bindToChange(name, regClass, onChange) {
            ($(document.body) as any).bind('scriptdatachange.' + regClass, function (e, s) {
                if (s == name) {
                    onChange();
                }
            });
        }

        export function triggerChange(name) {
            $(document.body).triggerHandler('scriptdatachange', [name]);
        }

        export function unbindFromChange(regClass) {
            $(document.body).unbind('scriptdatachange.' + regClass);
        }

        function syncLoadScript(url) {
            $.ajax({ async: false, cache: true, type: 'GET', url: url, data: null, dataType: 'script' });
        }

        function loadScriptAsync(url) {
            return RSVP.resolve().then(function () {
                Q.blockUI(null);
                return RSVP.resolve($.ajax({ async: true, cache: true, type: 'GET', url: url, data: null, dataType: 'script' }).always(function () {
                    Q.blockUndo();
                }));
            }, null);
        }

        function loadScriptData(name: string) {
            if (registered[name] == null) {
                throw new Error(Q.format('Script data {0} is not found in registered script list!', name));
            }

            name = name + '.js?' + registered[name];
            syncLoadScript(Q.resolveUrl('~/DynJS.axd/') + name);
        }

        function loadScriptDataAsync(name: string): RSVP.Thenable<any> {
            return RSVP.resolve().then(function () {
                if (registered[name] == null) {
                    throw new Error(Q.format('Script data {0} is not found in registered script list!', name));
                }

                name = name + '.js?' + registered[name];
                return loadScriptAsync(Q.resolveUrl('~/DynJS.axd/') + name);
            }, null);
        }

        export function ensure(name: string) {
            var data = loadedData[name];
            if (data == null) {
                loadScriptData(name);
            }

            data = loadedData[name];

            if (data == null)
                throw new Error(Q.format("Can't load script data: {0}!", name));

            return data;
        }

        export function ensureAsync(name: string): RSVP.Thenable<any> {
            return RSVP.resolve().then(function () {
                var data = loadedData[name];
                if (data != null) {
                    return RSVP.resolve(data);
                }

                return loadScriptDataAsync(name).then(function () {
                    data = loadedData[name];
                    if (data == null) {
                        throw new Error(Q.format("Can't load script data: {0}!", name));
                    }
                    return data;
                }, null);
            }, null);
        }

        export function reload(name: string) {
            if (registered[name] == null) {
                throw new Error(Q.format('Script data {0} is not found in registered script list!', name));
            }
            registered[name] = (new Date()).getTime().toString();
            loadScriptData(name);
            var data = loadedData[name];
            return data;
        }

        export function reloadAsync(name: string) {
            return RSVP.resolve().then(function () {
                if (registered[name] == null) {
                    throw new Error(Q.format('Script data {0} is not found in registered script list!', name));
                }
                registered[name] = (new Date()).getTime().toString();
                return loadScriptDataAsync(name).then(function () {
                    return loadedData[name];
                }, null);
            }, null);
        }

        export function canLoad(name: string) {
            return (loadedData[name] != null || registered[name] != null);
        }

        export function setRegisteredScripts(scripts) {
            registered = {};
            for (var k in scripts) {
                registered[k] = scripts[k].toString();
            }
        }

        export function set(name: string, value) {
            loadedData[name] = value;
            triggerChange(name);
        }
    }

    export function getRemoteData(key) {
        return ScriptData.ensure('RemoteData.' + key);
    }

    export function getRemoteDataAsync(key) {
        return ScriptData.ensureAsync('RemoteData.' + key);
    }

    export function getLookup<TItem>(key): Lookup<TItem> {
        return ScriptData.ensure('Lookup.' + key);
    }

    export function getLookupAsync<TItem>(key): RSVP.Thenable<Lookup<TItem>> {
        return ScriptData.ensureAsync('Lookup.' + key);
    }

    export function reloadLookup(key) {
        ScriptData.reload('Lookup.' + key);
    }

    export function reloadLookupAsync(key) {
        return ScriptData.reloadAsync('Lookup.' + key);
    }

    export function getColumns(key) {
        return ScriptData.ensure('Columns.' + key);
    }

    export function getColumnsAsync(key) {
        return ScriptData.ensureAsync('Columns.' + key);
    }

    export function getForm(key) {
        return ScriptData.ensure('Form.' + key);
    }

    export function getFormAsync(key) {
        return ScriptData.ensureAsync('Form.' + key);
    }

    export function getTemplate(key) {
        return ScriptData.ensure('Template.' + key);
    }

    export function getTemplateAsync(key) {
        return ScriptData.ensureAsync('Template.' + key);
    }

    export function canLoadScriptData(name) {
        return ScriptData.canLoad(name);
    }
}
